Resiliency: Add retry logic to stale ceps checks

derailed opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->

Resiliency: Add retry logic to stale CEPS checks

> NOTE: Depends on PR: #27614 

When the cilium agent starts it attempts to reconcile CEPS from loaded state. 
If a failure occurs no further attempts will take place, thus stale CEPS could remain on the cluster.

- Add retry logic to allow retries to take place should errors occur during CEPS reconciliation.

Fixes: #issue-number

```release-note
During startup, the agent attempts to clear out any obsolete CiliumEndpoints. 
Add retry logic to ensure this process is attempted more than once should errors occur during reconciliation.
```

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

christarazi commented on pkg/resiliency/types.go:
Could you split the addition of this package to a separate PR? It will make it easier to review and iterate on the package's API plus give us the flexibility in backporting. The usage code can still stay in this PR and it can be based off of the new PR.

derailed replied:
Good idea! will do...

derailed replied:
That said, I think some of these resiliency additions are likely going to be very invasive and likely not easy to back port...

derailed committed:
Resiliency: Add retry logic to track stale cep checks

When the cilium agent starts it attempts to reconcile ceps from loaded
state. If a failure occurs no further attempts will take place.

- Add retry logic to allow reconciliation to take place should errors
  occur during ceps gc.

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

tommyp1ckles commented on daemon/cmd/daemon_main.go:
```diff
@@ -1758,11 +1759,30 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+				Steps:    5,
+				Cap:      0,
+			}
+			err := wait.ExponentialBackoffWithContext(context.Background(), bo, func(ctx context.Context) (done bool, err error) {
```
One thing I'm wondering here is how we decide between using this approach vs pkg/controller (i.e. having a controller that runs until its error returns nil).

Also, another thought is that perhaps this might be a good case for the new "Job" type which supports retry until completion for OneShot jobs: https://github.com/cilium/cilium/blob/main/pkg/hive/job/job.go#L204

christarazi replied:
One part of the criteria is that older stable branches won't have "Job" type.

christarazi replied:
Controller pattern is the typical pattern in the Cilium code base. It already has logic to retry exponentially. It seems like it integrates well here in this case, so we might as well utilize it. 

With the controller pattern, it becomes easier to hold a queue of items that need to be processed and go through them one by one. If one fails, that item gets put back onto the queue. The controller will exit when the queue is empty. In this case, the item in the queue here is a stale CEP.

tommyp1ckles replied:
> One part of the criteria is that older stable branches won't have "Job" type.

Yup, good point 👍

thorn3r replied:
agreed, controller seems like the right fit here

christarazi replied:
In case you weren't aware @derailed, controller here means the infrastructure provided by `pkg/controller`.

derailed replied:
Somehow using a controller for this use case feels a bit heavy, the current impl already handles exp retries + retry limit. Possible with a controller of course but does not seems like the right use case for it in this instance imho.

christarazi replied:
While the difference between a controller and the current state of the PR is minimal in terms of an actual mechanism for retries, controllers have a bit more structure around them. Controllers have a status and that is aggregated and reported to (1) metrics and (2) `cilium status` output. If this controller is failing, the user could easily be informed by the controller infrastructure, while the current PR today would have to add code to come to parity. In this case, there's no added benefit to duplicating the code for the PR to come to parity with what `pkg/controller` already provides. Does this make sense?

derailed replied:
After discussing this with @christarazi, the current implementation is a better fit for this particular use case

tommyp1ckles commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -46,8 +50,9 @@ func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps localEndpointCache, cil
-					d.deleteCiliumEndpoint(ctx, ces.Namespace, cep.Name, nil, ciliumClient, eps,
-						enableCiliumEndpointSlice)
+					if err := d.deleteCiliumEndpoint(ctx, ces.Namespace, cep.Name, nil, ciliumClient, eps, enableCiliumEndpointSlice); err != nil {
+						errs = append(errs, err)
```
nice, I like the idea of still trying to attempt cleanup of other CEPs even if one fails.

aditighag reviewed (approved):
LGTM for my code owners (CODEOWNERS) change.

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -1775,11 +1776,30 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
+				return true, nil
+			})
+			if err != nil {
+				log.WithError(err).Error("Failed to clean up stale CEPs after retry")
```
```suggestion
					log.WithError(err).WithField(logfields.Attempt, x).Error("Failed to clean up stale CEPs")
					if resiliency.IsRetryable(err) {
						return false, err
					}
					return true, err
				}

				return true, nil
			})
			if err != nil {
				log.WithError(err).Error("Failed to clean up stale CEPs after multiple attempts")
```

I think we can add `attempt` here in the `WithFields` for extra context.

derailed replied:
Good point! I'll update accordingly.

christarazi commented on daemon/cmd/ciliumendpoints.go:
```diff
@@ -87,16 +93,16 @@ func (d *Daemon) deleteCiliumEndpoint(
-			} else {
-				log.WithError(err).WithFields(logrus.Fields{logfields.CEPName: cepName, logfields.K8sNamespace: cepNamespace}).
-					Error("Failed to get possibly stale ciliumendpoints from apiserver, skipping.")
+				return nil
```
We could so a bit of refactoring here and pull up the 

```
log.WithFields(logrus.Fields{
		logfields.CEPName:      cepName,
		logfields.K8sNamespace: cepNamespace,
```

a bit earlier and refactor the log statements which aren't DRY to use the logger fields from above.

derailed replied:
Thanks Chris! I saw that but figured it might be a bit out of scope for this pr??
I'll add the refactor in a separate commit.

derailed committed:
Refactor log lines

Dry up logging to encapsulate common fields.

Signed-off-by: Fernand Galiana <fernand.galiana@isovalent.com>

christarazi reviewed (approved):
LGTM. I think we need to drop the first commit about the resiliency package and it's good to go from my side
